### PR TITLE
EV charger energy management + car charger tests.

### PR DESCRIPTION
**Pull request by rstraczynski**

### CODE REVIEW

File: adapter/service/chargepoint/service.go [hunk 1]
```diff
@@ -40,6 +45,16 @@ type Controller interface {
 	ChargepointStateReport() (string, error)
 }
 
+// EnergyAwareController is an interface representing an actual car charger device with energy management capabilities.
+type EnergyAwareController interface {
```
Comment (boguslaw-wojcik) [inline adapter/service/chargepoint/service.go:49]: For me EnergyAware, besides bringing memories of high-ceremoniality of PHP-copycating-Java, seems misleading. Reporter for meter_elec is energy aware. This is controller which  is capable of adjusting charging mode.

Thesarus: Adjustable, Capable...

Maybe something like Adjustable(Charging)ModeController....

or... ChargingModeCapableController...

I don't know...

File: adapter/service/chargepoint/service.go [hunk 2]
```diff
@@ -64,6 +79,14 @@ type Service interface {
 	SendStateReport(force bool) (bool, error)
 	// SupportedStates returns states that are supported by the chargepoint.
 	SupportedStates() []string
+	// SupportsEnergyManagement returns true if chargepoint supports energy management.
+	SupportsEnergyManagement() bool
```
Comment (boguslaw-wojcik) [inline adapter/service/chargepoint/service.go:83]: Personally, I am really often guilty when it comes to YAGNI, so it's ironic, but maybe it should be just called SupportsChargingMode?

File: adapter/service/chargepoint/service.go [hunk 3]
```diff
@@ -260,3 +289,95 @@ func (s *service) SendStateReport(force bool) (bool, error) {
...
+
+	return true
+}
+
+func (s *service) SetChargingMode(mode string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if !s.SupportsEnergyManagement() {
+		return fmt.Errorf("%s: setting charging mode is is unsupported", s.Name())
```
Comment (boguslaw-wojcik) [inline adapter/service/chargepoint/service.go:311]: ..is is..

File: adapter/service/chargepoint/service.go [hunk 4]
```diff
@@ -260,3 +289,95 @@ func (s *service) SendStateReport(force bool) (bool, error) {
...
+	s.reportingCache.Reported(EvtChargingModeReport, "", mode)
+
+	return true, nil
+}
+
+// normalizeChargingMode normalizes provided charging mode. Returns true, when everything is fine.
+func (s *service) normalizeChargingMode(mode string) (string, bool) {
+	m := strings.ToLower(mode)
+
+	switch m {
```
Comment (boguslaw-wojcik) [inline adapter/service/chargepoint/service.go:376]: If you normalize against fixed list, then you should not allow to configure available charging modes when creating specification and just rely on automation that will add them when constructing service. 

If you allow an arbitraty list of charging modes to be set in specification, then you should normalize against properties in the specification.

The second solution is, I think, preferable, as it is more flexible. We have a special meaning and semi-locked charging modes, but we're not limiting ourselves to add more in future without a required change in the cliffhanger.

File: adapter/service/chargepoint/specification.go [hunk 5]
```diff
@@ -16,8 +16,9 @@ func Specification(
 	address string,
 	groups,
 	supportedStates []string,
+	supportedChargingModes []string, // Optional
```
Comment (boguslaw-wojcik) [inline adapter/service/chargepoint/specification.go:19]: See comment for `normalizeChargingMode` method.

File: adapter/service/chargepoint/specification.go [hunk 6]
```diff
@@ -95,3 +102,27 @@ func requiredInterfaces() []fimptype.Interface {
 		},
 	}
 }
+
+// energyManagementInterfaces returns optional energy management interfaces of the service.
```
Comment (boguslaw-wojcik) [inline adapter/service/chargepoint/specification.go:106]: If you change names for controller and support method, then change this one too.

**Review by boguslaw-wojcik — CHANGES_REQUESTED** [review]
You did not modify tasks to include reporting charging mode on startup and whenever it changes. Remember that adapter is not the only possible source or initiator of changes to the device.

File: test/mocks/adapter/service/chargepoint/energy_aware_controller_helper.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
+package mockedchargepoint
+
+func MockEnergyAwareController() *EnergyAwareController {
+	return &EnergyAwareController{}
+}
+
```
Comment (boguslaw-wojcik) [inline test/mocks/adapter/service/chargepoint/energy_aware_controller_helper.go:6]: Prolly you should copy all helpers from controller here as well.

File: adapter/thing/car_charger_test.go [hunk 8]
```diff
@@ -0,0 +1,589 @@
...
+}
+
+func TestTaskCarCharger(t *testing.T) { //nolint:paralleltest
+	s := &suite.Suite{
+		Cases: []*suite.Case{
+			{
+				Name: "Thermostat tasks",
+				Setup: taskCarCharger(
+					mockedchargepoint.MockController().
+						MockChargepointCableLockReport(true, nil, true).
```
Comment (boguslaw-wojcik) [inline adapter/thing/car_charger_test.go:460]: You did not test Charging state... Which is not implemented for tasks...

File: adapter/thing/car_charger_test.go [hunk 9]
```diff
@@ -0,0 +1,589 @@
...
+)
+
+var (
+	evChargerTestStates        = []string{"ready_to_charge", "charging", "error"}
+	evChargerChargingTestModes = []string{"normal", "slow"}
+
+	errTest = errors.New("oops")
+)
+
+func TestRouteCarCharger(t *testing.T) { //nolint:paralleltest
```
Comment (boguslaw-wojcik) [inline adapter/thing/car_charger_test.go:29]: You did not test out in routing scenarios for:
* "missing device" in adapter for each command 
* wrong format of input commands
While obviously this is not super important, it's easy to add, and we're struggling to increase cov here.

File: adapter/service/chargepoint/service.go (hunk 2 above)
Comment (rstraczynski) [reply adapter/service/chargepoint/service.go:83]: Yeah, can be renamed. Done.

File: adapter/service/chargepoint/service.go (hunk 3 above)
Comment (rstraczynski) [reply adapter/service/chargepoint/service.go:311]: Good catch! Fixed!

File: adapter/service/chargepoint/service.go [hunk 10]
```diff
@@ -260,3 +289,95 @@ func (s *service) SendStateReport(force bool) (bool, error) {
...
+		return fmt.Errorf("%s: unsupported charging mode: %s", s.Name(), mode)
+	}
+
+	controller, ok := s.controller.(EnergyAwareController)
+	if !ok {
+		return fmt.Errorf("%s: setting charging mode is unsupported", s.Name())
+	}
+
+	if err := controller.SetChargepointChargingMode(normalizedMode); err != nil {
+		return fmt.Errorf("%s: failed to retrieve extended meter report: %w", s.Name(), err)
```
[the remark this answers is not in the document]
Comment (rstraczynski) [reply adapter/service/chargepoint/service.go:325]: Right, fixed.

File: adapter/service/chargepoint/service.go (hunk 4 above)
Comment (rstraczynski) [reply adapter/service/chargepoint/service.go:376]: You're absolutely right, that's how I wanted it to work at the first place (I mean the second option) - not sure why I've implemented it that way 😄 
Fixed.

File: adapter/service/chargepoint/specification.go (hunk 5 above)
Comment (rstraczynski) [reply adapter/service/chargepoint/specification.go:19]: Resolved.

File: test/mocks/adapter/service/chargepoint/energy_aware_controller_helper.go (hunk 7 above)
Comment (rstraczynski) [reply test/mocks/adapter/service/chargepoint/energy_aware_controller_helper.go:6]: Done.

File: adapter/service/chargepoint/specification.go (hunk 6 above)
Comment (rstraczynski) [reply adapter/service/chargepoint/specification.go:106]: Done.

File: adapter/thing/car_charger_test.go (hunk 8 above)
Comment (rstraczynski) [reply adapter/thing/car_charger_test.go:460]: At first I rejected the idea of charging mode reporting, but then I looked at current session report :) Fixed!

File: adapter/thing/car_charger_test.go (hunk 9 above)
Comment (rstraczynski) [reply adapter/thing/car_charger_test.go:29]: OK, done.

**Review by boguslaw-wojcik — APPROVED** [review]

**Pull request merged**